File: src/parser/parser.py
from .lexer import Token, TokenType
from ..utils.errors import Error
from typing import List

class Parser:
    def __init__(self, tokens, path):
        self.program = {
            "Program": {
                "body": []
            }
        }
        self.tokens: List[Token] = tokens
        self.path = path
    
    def get_type(self, type: TokenType):
        return str(type).split(".")[1]

    def advance(self):
        return self.tokens.pop(0)
        
    def push(self, content):
        return self.program["Program"]["body"].append(content)
    
    def expect(self, type):
        tok = self.tokens[0]
        if not tok or tok.type != type:
            Error(self.path, tok.location[0], tok.location[1], "Unexpected token", f"Expected token of type `{self.get_type(type)}`, but found `{self.get_type(tok.type)}`", False)

        return tok

    def generateAST(self): 
        while self.tokens[0].type != TokenType.EOF:
            self.push(self.parse_stmt())

        return self.program
    
    def parse_stmt(self):
        return self.parse_expr()
    
    def parse_expr(self):
        return self.parse_additive_expr()
    
    def parse_additive_expr(self):
        # Parses additive expressions (addition and subtraction)
        left = self.parse_multiplicative_expr()
        
        while self.tokens[0].value in ["+", "-"]:
            op = self.advance().value
            right = self.parse_multiplicative_expr()
            left = {
                "kind": "BinaryExpr",
                "left": left,
                "right": right,
                "op": op
            }
            
        return left
    
    def parse_multiplicative_expr(self):
        # Parses multiplicative expressions (multiplication, division and modulus)
        left = self.parse_primary_expr()
        
        while self.tokens[0].value in ["*", "/", "%"]:
            op = self.advance().value
            right = self.parse_primary_expr()
            left = {
                "kind": "BinaryExpr",
                "left": left,
                "right": right,
                "op": op
            }
            
        return left

# Order of precedence
    # Assignment expression
    # Member expression
    # Function call
    # Logical expression
    # Comparison expression
    # Additive expression
    # Multiplicative expression
    # Unary expression
    # Primary expression

    def parse_primary_expr(self):
        # Parses primary expressions like integer, literals and keywords
        match self.tokens[0].type:
            case TokenType.Identifier:
                return {
                    "kind": "Identifier",
                    "value": self.advance().value
                }
            
            case TokenType.Null:
                self.advance()
                return {
                    "kind": "NullLiteral",
                    "value": "null"
                }

            case TokenType.Integer:
                return {
                    "kind": "NumericLiteral",
                    "value": int(self.advance().value)
                }
            
            case TokenType.Float:
                return {
                    "kind": "NumericLiteral",
                    "value": float(self.advance().value)
                }
            
            case TokenType.OpenParen:
                self.advance() # Open Paren
                value = self.parse_expr() # Expression
                self.expect(TokenType.CloseParen)
                self.advance() # Close Paren
                return value
            
            case _:
                err = self.advance()
                Error(self.path, err.location[0], err.location[1], "Unexpected token", "Unexpected token found:", err.value)

__all__ = [
    "Parser"
]

if __name__ == "__main__":
    import os
    raise Exception(f"`{os.path.basename(__file__)} can\'t be ran as `{__name__}`")

File: src/parser/parser.js
const { Stmt, Program, Expr, BinaryExpr, Identifier, NumericLiteral, NullLiteral } = require("./ast")
const { Token, TokenType } = require("./lexer/tokens")
const { Error } = require("../utils/errors")

class Parser {
    constructor(path, tokens) {
        this.program = new Program()
        
        this.path = path
        this.tokens = tokens
    }
    
    // Gets key of object from value 
    getType(type) {        
        return Object.keys(TokenType).find(key => TokenType[key] === type)
    }

    advance() {
        return this.tokens.shift()
    }

    push(content) {
        return this.program.body.push(content)
    }
        
    expect(type) {
        const tok = this.tokens[0]

        if (!tok || tok.type != type) {
            new Error(this.path, tok.loc.line, tok.loc.col, "Unexpected token", `Expected token of type \`${this.getType(type)}\`, but found \`${this.getType(tok.type)}\``)
        }

        return tok
    }

    generateAST() {
        while (this.tokens[0].type != TokenType.EOF) {
            this.push(this.parse_stmt())
        }

        return this.program
    }

    parse_stmt() {
        return this.parse_expr()
    }

    parse_expr() {
        return this.parse_additive_expr()
    }

    // Parses additive expressions (addition and subtraction)
    parse_additive_expr() {
        let left = this.parse_multiplicative_expr() 

        while (["+", "-"].some((k) => this.tokens[0].value == k)) {
            const op = this.advance().value
            const right = this.parse_multiplicative_expr()

            left = new BinaryExpr(left, right, op)
        }

        return left
    }
    
    // Parses multiplicative expressions (multiplication, division and modulus)
    parse_multiplicative_expr() {
        let left = this.parse_primary_expr()

        while (["*", "/", "%"].some((k) => this.tokens[0].value == k)) {
            const op = this.advance().value
            const right = this.parse_primary_expr()

            left = new BinaryExpr(left, right, op)
        }

        return left
    }

    /* Order of precedence
        * Assignment expression
        * Member expression
        * Function call
        * Logical expression
        * Comparison expression
        * Additive expression
        * Multiplicative expression
        * Unary expression
        * Primary expression
    */

    // Parses primary expressions like integer, literals and keywords that map to built-in types
    parse_primary_expr() {
        switch (this.tokens[0].type) {

            case TokenType.Identifier:
                return new Identifier(this.advance().value)

            case TokenType.Null:
                this.advance()
                return new NullLiteral()

            case TokenType.Integer:
                return new NumericLiteral(parseInt(this.advance().value))

            case TokenType.Float:
                return new NumericLiteral(parseFloat(this.advance().value))

            case TokenType.OpenParen:
                this.advance() // Eat the `(`
                const value = this.parse_expr() // Parse the expression inside
                this.expect(TokenType.CloseParen)
                this.advance() // Eat the `)`

                return value

            default:
                const error = this.advance()
                new Error(this.path, error.loc.line, error.loc.col, "Unexpected token", `Unexpected token: \`${error.value}\`.`) 
        }
    }
}

module.exports = {
    Parser
}

if (require.main === module) {
    const process = require("process")
    const path = require("path")

    console.error(`ERROR: \`${path.parse(__filename).name}\` file can\'t be ran as main.`)
    process.exit(1)
}